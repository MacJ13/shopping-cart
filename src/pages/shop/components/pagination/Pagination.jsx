/* eslint-disable react/prop-types */
import { useSearchParams } from "react-router-dom";
import { isNumber, paginate } from "../../../../utils/helpers";
import style from "./Pagination.module.scss";

const Pagination = ({ totalPages, currentPage }) => {
  let [searchParams, setSearchParams] = useSearchParams();

  const onClickPageButton = (page) => {
    searchParams.set("page", page);
    searchParams.sort();
    setSearchParams(searchParams);
  };

  const pages = paginate(currentPage, totalPages);

  if (currentPage === 1 && currentPage === totalPages) return null;

  return (
    <div className={style.pagination}>
      <div className={style.bar}>
        {pages.map((page, i) => {
          return isNumber(page) ? (
            <div key={page} className={style.btn}>
              <button
                className={page === currentPage ? style.active : ""}
                onClick={() => {
                  onClickPageButton(page);
                }}
              >
                {page}
              </button>
            </div>
          ) : (
            <div
              // data-id={"dot" + page + "" + i}
              key={"dot" + page + "" + i}
              className={style.dot}
            >
              {page}
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default Pagination;
