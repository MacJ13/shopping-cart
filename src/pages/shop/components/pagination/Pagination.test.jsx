import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import Pagination from "./Pagination";
import { BrowserRouter } from "react-router-dom";

describe("Pagination component", () => {
  it("renders number pages", () => {
    const { container } = render(
      <BrowserRouter>
        <Pagination currentPage={5} totalPages={10} />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("should renders simply page buttons with current page equals 5 and total pages are 10 ", () => {
    render(
      <BrowserRouter>
        <Pagination currentPage={5} totalPages={10} />
      </BrowserRouter>
    );

    const buttons = screen.getAllByRole("button");
    const dots = screen.getAllByText(/.../i);

    expect(dots.length).toBe(1);
    expect(dots[0]).toBeInTheDocument();

    expect(buttons[0].textContent).toMatch(/1/i);
    expect(buttons[1].textContent).toMatch(/2/i);
    expect(buttons[2].textContent).toMatch(/3/i);
    expect(buttons[3].textContent).toMatch(/4/i);

    expect(buttons[buttons.length - 3].textContent).toMatch(/7/i);
    expect(buttons[buttons.length - 2].textContent).toMatch(/8/i);
    expect(buttons[buttons.length - 1].textContent).toMatch(/10/i);
  });

  it("should renders simply page buttons with current page equals 2 and total pages are 4 ", () => {
    render(
      <BrowserRouter>
        <Pagination currentPage={2} totalPages={4} />
      </BrowserRouter>
    );

    const buttons = screen.getAllByRole("button");
    const dot = screen.queryByText(/.../i);

    expect(dot).not.toBeInTheDocument();

    expect(buttons[0].textContent).toMatch(/1/i);
    expect(buttons[1].textContent).toMatch(/2/i);
    expect(buttons[2].textContent).toMatch(/3/i);
    expect(buttons[3].textContent).toMatch(/4/i);
  });

  it("should renders page buttons with current page equals 8 and total pages are 16 ", () => {
    render(
      <BrowserRouter>
        <Pagination currentPage={8} totalPages={16} />
      </BrowserRouter>
    );

    const dots = screen.getAllByText(/.../i);

    expect(dots.length).toBe(2);
    expect(dots[0]).toBeInTheDocument();
    expect(dots[1]).toBeInTheDocument();
  });

  it("should not renders nothing", () => {
    render(
      <BrowserRouter>
        <Pagination currentPage={1} totalPages={1} />
      </BrowserRouter>
    );

    const button = screen.queryByRole("button");

    expect(button).not.toBeInTheDocument();
  });
});
