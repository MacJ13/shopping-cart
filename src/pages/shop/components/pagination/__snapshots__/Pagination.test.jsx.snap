// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination component > renders number pages 1`] = `
<div>
  <div
    class="_pagination_e4c38a"
  >
    <div
      class="_bar_e4c38a"
    >
      <div
        class="_btn_e4c38a"
      >
        <button
          class=""
        >
          1
        </button>
      </div>
      <div
        class="_btn_e4c38a"
      >
        <button
          class=""
        >
          2
        </button>
      </div>
      <div
        class="_btn_e4c38a"
      >
        <button
          class=""
        >
          3
        </button>
      </div>
      <div
        class="_btn_e4c38a"
      >
        <button
          class=""
        >
          4
        </button>
      </div>
      <div
        class="_btn_e4c38a"
      >
        <button
          class="_active_e4c38a"
        >
          5
        </button>
      </div>
      <div
        class="_btn_e4c38a"
      >
        <button
          class=""
        >
          6
        </button>
      </div>
      <div
        class="_btn_e4c38a"
      >
        <button
          class=""
        >
          7
        </button>
      </div>
      <div
        class="_btn_e4c38a"
      >
        <button
          class=""
        >
          8
        </button>
      </div>
      <div
        class="_dot_e4c38a"
      >
        ...
      </div>
      <div
        class="_btn_e4c38a"
      >
        <button
          class=""
        >
          10
        </button>
      </div>
    </div>
  </div>
</div>
`;
